lib/src/stream_queue.dart

```diff
@@ -407,6 +458,84 @@ class _StreamQueue<T> extends StreamQueue<T> {
   }
 }
 
+/// A [StreamQueue] cloned from a parent by [StreamQueue.startTransactions].
+///
+/// This queue may be [accept]ed, moving the parent to the same event as this,
+/// or [reject]ed, marking this as finished without modifying the parent.
+class TransactionStreamQueue<T> extends _StreamQueue<T> {
+  /// The request that created this queue.
+  final _TransactionRequest<T> _request;
+
+  /// Whether [accept] has been called.
+  var _accepted = false;
+
+  /// Whether [reject] has been called.
+  var _rejected = false;
+
+  /// The total number of events received by this queue, including events that
+  /// haven't yet been consumed by requests.
+  ///
+  /// This is used to fast-forward the parent queue if this transaction is
+  /// accepted.
+  var _eventsReceived = 0;
+
+  TransactionStreamQueue._(Stream<T> sourceStream, this._request)
+      : super(sourceStream);
+
+  /// Updates the parent [StreamQueue] to match this transaction's state.
+  ///
+  /// Further requests on this queue will be completed as though the underlying
+  /// stream had been closed, as though [cancel] were called with `immediate:
+  /// true`. Other transactions in the same batch are automatically rejected.
+  ///
+  /// Throws a [StateError] if [accept] or [reject] has already been called, or
+  /// if there are pending requests on this queue.
+  void accept() {
+    _assertActive();
+
+    if (_requestQueue.isNotEmpty) {
+      throw new StateError(
+          "A transaction with pending requests can't be accepted.");
```

Comment:
I went with the "just throw" solution because I didn't think it was really clear which of the several possible behaviors was the most obvious or most useful, and I couldn't come up with any really compelling use-cases that weren't supportable in other ways. If we throw for now it leaves us open to choose a different behavior in the future if it becomes clear that one is correct.